[update] MIP-0 : Introduce status terms and editor role (#32)
movementlabsxyz/MIP: README.md

## Files and numbering
Each MIP, MD or MG is stored in a separate subdirectory with the a name `mip-<number>`, `md-<number>` or `mg-<number>`. The subdirectory contains a `README.md` file that describes the MIP, MD, or MG. All assets related to the MIP, MD or MG are stored in the same subdirectory.

MIPs start as Drafts. MIPs obtain a number during the review process described in [Status Terms](#status-terms). 

PRs that don't introduce a new MIP are also accepted. 
- MIPs can be updated. PRs that update a MIP should state so in the PR title via `[Update] .... `. 

There may be gaps in the numbering, as MIPs get rejected and removed.

## Status Terms
An MIP is proposed through a PR. Each MIP-introducing PR should have a status in the name in the form `[Status] ...`.

An MIP should at all times have one of the following statuses:
- **Draft** - (set by author) An MIP that is open for consideration. (It does not yet hold an MIP number)

> - **Review** - (set by author) The MIP is under peer review. The MIP should receive an **MIP number** and the MIP should be registered in the MIP overview file.

Reviewer: @apenzk I like the statuses. Could you add more details about how the MIP number is assigned and registered? Perhaps there should be more about the editor role as well, if that's a new addition to our MIP process?

Author: I have removed the mention of the overview file, as it will be included by this [[update]](https://github.com/movementlabsxyz/MIP/pull/31/files)

i have added a note to clarify where the MIP number comes from. Essentially it will be fixed by the OVERVIEW file which comes with [this PR](https://github.com/movementlabsxyz/MIP/pull/31/files) from Franck.

Author: > It may be helpful to define or introduce the editor role as well, if that's a new addition to our MIP process?

done.

> - **Accepted** - (set by editor) An MIP that has been accepted. All technical changes requested have been addressed by the author. There may be additional non-technical changes requested by the MIP editor.

Reviewer: One issue that may arise is: suppose an editor has a technical change that the author doesn't agree with. Is there some mechanism that we could add, to resolve such disagreements or disputes?

Author: (i have addressed now the second para in the text)

Reviewer: > > suppose an editor has a technical change that the author doesn't agree with. Is there some mechanism that we could add, to resolve such disagreements or disputes
> 
> How is this handled with code PRs?

For PRs, we have a minimum number of approvals needed (3 for MIPs, 2 for PRs to `aptos-core` and 1 for PRs to `movement`).

What about this solution: If an editor requests a change from an author, either the author can make the changes (tacitly approving the editorial feedback), or:

If the author doesn't agree with the changes, then the editor can mandate that the author implements the change suggestion by getting 2 upvotes on their discussion comment with the change requests (assuming change requests will be made in the Discussion section of the MIP PR)?

Author: sounds good. added this. thx!

## Files and numbering
Each MIP, MD or MG is stored in a separate subdirectory with the a name `mip-<number>`, `md-<number>` or `mg-<number>`. The subdirectory contains a `README.md` file that describes the MIP, MD, or MG. All assets related to the MIP, MD or MG are stored in the same subdirectory.

> MIPs start as Drafts. MIPs obtain a number during the review process. The number is the highest number in the MIP and MD `overview.md` file plus 1.

Author: ```suggestion
MIPs start as Drafts. MIPs obtain a number during the review process. The number is the highest number in the MIP and MD `overview.md` file plus 1. The editor can be requested for support on this.
```

PRs that don't introduce a new MIP are also accepted. 
- MIPs can be updated. PRs that update a MIP should state so in the PR title via `[Update] .... `. 

There may be gaps in the numbering, as MIPs get rejected and removed.

## Status Terms

> An MIP is proposed through a PR. Each MIP-introducing PR should have a status in the name in the form `[Status] ...`.

Reviewer 2: Can you provide a justification for this as opposed to labels?

Author: There is no good reason for this as opposed to labels. I just compared with 
https://github.com/ethereum/EIPs/pulls and preferred it over not having anything. Would you prefer labels?

Reviewer 2: I think I would. If there is another opinion that supports mine, I would say we go with that.

An MIP should at all times have one of the following statuses:
- **Draft** - (set by author) An MIP that is open for consideration. (It does not yet hold an MIP number)
- **Review** - (set by author) The MIP is under peer review. The MIP should receive an **MIP number**, according to the rules described in the [Files and numbering](#files-and-numbering) section.

> >[!Note]

Reviewer 2: Good.

Author: que?

Reviewer 2: This is a good thing to note.

> In case the editors are not available for an unacceptable long period of time, a reviewer should assume the role of the editor interim. 

> 

Reviewer 2: I might note in this very README that a big motivation for the role of the editor is the native readability of this format. We won't likely have something that re-renders the content herein for a while, so for people new to the format--particularly external users--the quality the editor ensures is important.

Author: ok done at the bottom section

After acceptance the MIP is merged into `main` and the branch should be deleted.

Addtionally, the following statuses are used for MIPs that are not actively being worked on:
- **Stagnant** - an MIP that has not been updated for 6 months.

> - **Withdrawn** - an MIP that has not been updated for 6 months.

Reviewer 2: I think the description on Withdrawn needs to be updated here.

Author: done.

> - **Withdrawn** - an MIP that has not been updated for 6 months.

Author: ```suggestion
- **Withdrawn** - an MIP that has not been revoked.
```

Finally, MIPs can also be updated
- **Update** - (set by author) An MIP is being updated. The titel should list the MIP number, e.g. `[Update] MIP-0 ...`.

## Editor

> Currently the editors are [@apenzk](https://github.com/apenzk). 

Author: ```suggestion
The motivation for the role of the editor is to ensure the readability and easy access of content, until further means, such as automatic rendering becomes available.

Currently the editors are [@apenzk](https://github.com/apenzk). 
```

The editor is responsible for the final review of the MIPs. The editor is responsible for the following:
- Ensures a high quality of the MIPs, e.g. checking language while reviewing.
- Removes content from the MIPs that is commented out. (<!- ->)
- Ensures the MIP numbering is correct, the MIP has been added to [OVERVIEW.md](./OVERVIEW.md)
- Ensures the MIP is in the correct status. 

> - Ensures the authors have added themselves to [CODEOWNERS](./.github/CODEOWNERS).

Reviewer 2: I would add the note here that this is about shifting some work for the editor going forward. In particular, we want to add the authors to the CODEOWNERS such that they will be tagged on future revisions for a particular MD or MIP, e.g., errata.

Author: PR 33 has been merged and we reference here now to the relevant section

- **Review** - (set by author) The MIP is under peer review. The MIP should receive an **MIP number**, determined by the author.

>[!Note]
> Currently the author has to understand from the PRs what the latest MIP number is. This is suboptimal and will be fixed by a later PR. 

- **Accepted** - (set by editor) An MIP that has been accepted. All technical changes requested have been addressed by the author. There may be additional non-technical changes requested by the MIP editor.

>[!Note]
> In case the editors are not available for an unacceptable long period of time, a reviewer should assume the role of the editor interim. 

After acceptance the MIP is merged into `main` and the branch should be deleted.

Addtionally, the following statuses are used for MIPs that are not actively being worked on:
- **Stagnant** - an MIP that has not been updated for 6 months.
- **Withdrawn** - an MIP that has not been updated for 6 months.

Finally, MIPs can also be updated
- **Update** - (set by author) An MIP is being updated. The titel should list the MIP number, e.g. `[Update] MIP-0 ...`.

## Editor

> Currently the editors are [@apenzk](https://github.com/apenzk), [@andyjsbell](https://github.com/andyjsbell), [@l-monninger](https://github.com/l-monninger). Volunteers are welcome.

Author: ```suggestion
Currently the editors are [@apenzk](https://github.com/apenzk). Volunteers are welcome.
```

> Currently the editors are [@apenzk](https://github.com/apenzk). Volunteers are welcome.

Author: ```suggestion
Currently the editors are [@apenzk](https://github.com/apenzk). 
```

> Currently the editors are [@apenzk](https://github.com/apenzk). Volunteers are welcome.

Author: the intention was to invite company internal volunteers as opposed to external reviewers. i have removed the sentence as it was unclear.

> Currently the editors are [@apenzk](https://github.com/apenzk). Volunteers are welcome.

Author: the editor can be bypassed as the rule remains that 3 approvals suffice. in which case the editor would have to apply [updates] if he is not satisfied. it is a non-blocking role. I have clarified this in the text.

> Currently the editors are [@apenzk](https://github.com/apenzk). Volunteers are welcome.

Author: the point of [accepted] is that it gives the editor the chance to clean up the MIPs, ensure that MIP creators are added as code owners.

> Currently the editors are [@apenzk](https://github.com/apenzk). Volunteers are welcome.

Reviewer: I'm not convinced anymore that the editor role and list are needed. If we just consider merged MIP PRs to be **Accepted** then the list of merged PRs can be our Accepted "list". Draft PRs are drafts and open PRs are in review or stale. 

The weekly MIP ceremonies can help us develop a discipline as a team to focus on relevant MIPs. Or in general if an MIP is falling by the wayside, we can ping other team members to review it. I don't think the editor role or a separate list are needed for this functionality.

> Currently the editors are [@apenzk](https://github.com/apenzk). Volunteers are welcome.

Author: i have reduced the role of editor as well as removed the accepted stage. still think that the drafts are important and should be easy to oversee and which they currently are not.. 

here is the example i draw from at Ethereum https://eips.ethereum.org/core

> Currently the editors are [@apenzk](https://github.com/apenzk). Volunteers are welcome.

Reviewer: It makes sense in that case, because they're using a separate website apart from GitHub. I think if this level of extra administration is not strictly necessary, then it should be avoided.

> Currently the editors are [@apenzk](https://github.com/apenzk). Volunteers are welcome.

Author: thx for the feedback though.. it really helps to understand what is actually needed.

> Currently the editors are [@apenzk](https://github.com/apenzk). Volunteers are welcome.

Reviewer 2: @andygolay I would argue that because we will not be re-rerendering this content for a while, we should be taking extra care about its presentation natively.

> Currently the editors are [@apenzk](https://github.com/apenzk). Volunteers are welcome.

Reviewer: Agreed, with proper readability and quality standards in place that we're all aware of, it can be easy to implement.
